design/metadata_kv/index.md

```diff
@@ -0,0 +1,187 @@
+
+### Graveler Metadata - Commits, Tags, Repositories
+
+These are simpler entities - commits and tags are immutable and could potentially be stored on the object store and not in the key/value store (they are cheap either way, so it doesn't make much of a difference).
+Repositories and tags, are also returned in lexicographical order, which map well to the suggested abstraction. Commits are usually returned using parent traversal, so no scanning takes place anyway. 
+
+There are no special concurrency requirements for these entities, apart for last-write-wins which is already the case for all modern stores.
+
+### Graveler Metadata - Branches and Staged Writes
+
+This is where concurrency control gets interesting, and where lakeFS is expected to provide a **correct** system whose semantics are well understood (lakeFS currently [falls short](https://github.com/treeverse/lakeFS/issues/2405) in that regard).
+
+Concurrency is more of an issue here because of how a commit works: when a commit starts, it scans the currently staged changes, applies them to the current commit pointed to by the branch, updating the branch reference and removing the staged changes it applied.
+
+Getting this right means we have to take care of the following:
+
+1. Ensure all staged changes that finished successfully before the commit started are applied as part of the commit (causality)
+1. Ensure acknowledged writes end up either in a resulting commit, or staged to be committed (no lost writes)
+
+To do this, we will employ 2 mechanisms:
+
+1. [Optimistic Concurrency Control](https://en.wikipedia.org/wiki/Optimistic_concurrency_control) on the branch pointer using `SetIf()`
+1. Reliance on write [idempotency](https://en.wikipedia.org/wiki/Idempotence) provided by Graveler (i.e., Writing the same exact entry, with the same identity - will not appear as a change)
+
+This is what the proposed implementation will look like:
+
+#### Committer flow
+
+We add an additional field to each `Branch` object: In addition to the existing `staging_token`, we add an array of strings named `sealed_tokens`.
+
+1. get branch, find current `staging_token`
+1. use `SetIf()` to update the branch (if not modified by another process): push existing `staging_token` into `sealed_tokens`, set new uuid as `staging_token`.
+1. take the list of sealed tokens, and using the [`CombinedIterator()`](https://github.com/treeverse/lakeFS/blob/master/pkg/graveler/combined_iterator.go#L11), turn them into a single iterator to be applied on top of the existing commit
+1. Once the commit has been persisted (metaranges and ranges stored in object store, commit itself stored to KV using `Set()`), perform another `SetIf()` that updates the branch again: replacing its commit ID with the new value, and clearing `sealed_tokens`, as these have materialized into the new commit.
+1. If `SetIf()` fails, this means another commit is happenning/has happened on the same branch. Can either retry or let the caller know.
```

Comment:
What is this `SetIf` on?  AFAICT just `staging_token` or just `commit_id` are not enough to provide ordering in the face of concurrent commits.  It might be if the entire value is unchanged, or possibly `staging_token` and `sealed_tokens`.

Reply:
This was indeed unclear - made it explicit that `SetIf()` is actually a CAS operation on the entire branch (`staging_token`, `sealed_tokens` and `commit_id`).